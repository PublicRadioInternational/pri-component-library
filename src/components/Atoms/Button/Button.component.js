/**
 * @file Button.component.js
 * Exports a classic button component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.css';
import Icon from '../Icon/Icon.component';

/**
 * Component that renders a link, or a button with a click handler.
 */
const Button = props => {
  const { url, onClick, className, children, color, icon, small } = props;
  // Generate a class name based on the color.
  const buttonClass = `btn${color}`;
  const buttonMobileClass = `btnMobile${color}`;

  // If a URL is provided, this button is simply a link.
  if (url) {
    return (
      <a
        href={url}
        className={`${
          small ? styles[buttonMobileClass] : styles[buttonClass]
        } ${className && className}`}
        onClick={onClick}
      >
        {icon && <Icon svg={icon} inline />}
        <span
          className={`${styles.textLabel} ${small && styles.textLabelMobile}`}
        >
          {children}
        </span>
      </a>
    );
  }

  // No URL was specified, this button is not a link. Return a button instead.
  return (
    <button
      type="button"
      className={`${
        small ? styles[buttonMobileClass] : styles[buttonClass]
      } ${className && className}`}
      onClick={onClick}
      aria-expanded={props['aria-expanded']}
      aria-label={props['aria-label']}
      aria-haspopup={props['aria-haspopup']}
    >
      {icon ? <Icon svg={icon} inline /> : null}
      {children}
    </button>
  );
};

Button.propTypes = {
  url: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  color: PropTypes.oneOf(['Orange', 'White']),
  className: PropTypes.string,
  'aria-expanded': PropTypes.bool,
  'aria-label': PropTypes.string,
  'aria-haspopup': PropTypes.bool,
  icon: PropTypes.string,
  small: PropTypes.bool
};

Button.defaultProps = {
  url: null,
  color: 'White',
  className: null,
  children: null,
  onClick: () => {},
  'aria-expanded': false,
  'aria-label': null,
  'aria-haspopup': false,
  icon: null,
  small: false
};

export default Button;
