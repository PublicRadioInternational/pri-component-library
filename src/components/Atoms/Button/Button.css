/* import colors... */
@value colors: "../../00_global/colors.css";
@value white, whiteDark, grayLighter, gray, grayLight, grayDark, grayDarker, orange, orangeDark, orangeDarker, blue, blueMed, blueDark, yellow from colors;
/* import fonts... */
@value fonts: "../../00_global/fonts.css";
@value copy from fonts;
/* import breakpoints */
@value header as bp-header from "../../00_global/breakpoints.css";

/* Generic */
@define-mixin btn {
  background-color: whiteDark;
  border: 1px solid grayLighter;
  border-radius: 0.25rem;
  color: grayDark;
  cursor: pointer;
  display: inline-block;
  font: 700 1rem copy;
  letter-spacing: 0.05rem;
  line-height: 2.25;
  margin: 0;
  padding: 0 0.85rem;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: color 0.15s ease-in-out, background-color 0.15s ease-in-out,
    border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
}

@define-mixin btnWhiteHover {
  background-color: white;
}

@define-mixin btnWhiteActive {
  background-color: grayDark;
  color: white;
}

/* Inherits button always */
.btn,
.btnWhite,
.btnOrange,
.btnDark,
.btnBlue {
  @mixin btn;
}

.btn:hover,
.btnWhite:hover {
  @mixin btnWhiteHover;
}

.btn:active,
.btn[aria-expanded='true'],
.btnWhite:active,
.btnWhite[aria-expanded='true'] {
  @mixin btnWhiteActive;
}

/* Colors */
@define-mixin btnOrange {
  background-color: orange;
  border-color: yellow;
  color: white;
}

@define-mixin btnOrangeHover {
  background-color: orangeDark;
  color: white;
}

@define-mixin btnBlue {
  background-color: blue;
  border: 1px solid blueMed;
  color: white;
}

@define-mixin btnBlueHover {
  background-color: blueMed;
  color: white;
}

.btnOrange {
  @mixin btnOrange;
}

.btnOrange:hover,
.btnOrange:active,
.btnOrange[aria-expanded='true'] {
  @mixin btnOrangeHover;
}

/* Colors */
@define-mixin btnDark {
  background-color: grayDarker;
  border-color: transparent;
  color: white;
}

@define-mixin btnDarkHover {
  background-color: orangeDarker;
  color: white;
}

.btnDark {
  @mixin btnDark;
}

.btnDark:hover,
.btnDark:active,
.btnDark[aria-expanded='true'] {
  @mixin btnDarkHover;
}

/* Colors */
@define-mixin btnBlue {
  background-color: blue;
  border-color: transparent;
  color: white;
}

@define-mixin btnBlueHover {
  background-color: orange;
  color: white;
}

.btnBlue {
  @mixin btnBlue;
}

.btnBlue:hover,
.btnBlue:active,
.btnBlue[aria-expanded='true'] {
  @mixin btnBlueHover;
}

/* Mobile Icon versions */
.btnMobileWhite,
.btnMobileDark {
  color: grayLight;
}

.btnMobileWhite:hover {
  color: grayDark;
}

.btnMobileOrange {
  color: orange;
}

.btnMobileBlue {
  color: blue;
}

@media bp-header {
  .btnMobileWhite,
  .btnMobileOrange,
  .btnMobileDark,
  .btnMobileBlue {
    @mixin btn;
  }

  .btnMobileWhite:hover {
    @mixin btnWhiteHover;
  }

  .btnMobileWhite:active,
  .btnMobileWhite[aria-expanded='true'] {
    @mixin btnWhiteActive;
  }

  .btnMobileOrange {
    @mixin btnOrange;
  }

  .btnMobileBlue {
    @mixin btnBlue;
  }

  .btnMobileOrange:hover,
  .btnMobileOrange:active,
  .btnMobileOrange[aria-expanded='true'] {
    @mixin btnOrangeHover;
  }

  .btnMobileDark {
    @mixin btnDark;
  }

  .btnMobileDark:hover,
  .btnMobileDark:active,
  .btnMobileDark[aria-expanded='true'] {
    @mixin btnDarkHover;
  }

  .btnMobileBlue {
    @mixin btnBlue;
  }

  .btnMobileBlue:hover,
  .btnMobileBlue:active,
  .btnMobileBlue[aria-expanded='true'] {
    @mixin btnBlueHover;
  }
}

/* Input Specific */
.inputWrap {
  position: relative;
}

svg.inputIcon {
  cursor: pointer;
  left: 10px;
  margin-top: -9px;
  position: absolute;
  top: 50%;
}

.inputBtn {
  @composes: btnWhite;
}

.inputWithIcon {
  padding: 0.375rem 0.85rem 0.375rem 2rem;
}

.inputWithIcon:active {
  background-color: grayLighter;
  color: grayDark;
}

/* Buttons in children */
@define-mixin btnSmall {
  font: 0.875rem 'Raleway', sans-serif;
  letter-spacing: 0;
  padding: 0.5rem 0.7rem;
  text-transform: none;
}

.btnWrap input[type='submit'] {
  @mixin btn;
  @mixin btnOrange;
  @mixin btnSmall;
}

.btnWrap input[type='submit']:hover {
  @mixin btnOrangeHover;
}
