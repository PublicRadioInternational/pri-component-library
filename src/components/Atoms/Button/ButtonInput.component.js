/**
 * @file ButtonInput.component.js
 * Exports an input button component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.css';
import Icon from '../Icon/Icon.component';

/**
 * Component that renders an input button with a click handler.
 */
const ButtonInput = props => {
  const { onClick, className, value, color, svg, small } = props;
  // Generate a class name based on the color.
  const buttonClass = `btn${color}`;
  const buttonMobileClass = `btnMobile${color}`;

  return (
    <div className={styles.inputWrap}>
      {svg ? <Icon svg={svg} inline aria-hidden className="inputIcon" /> : null}
      <input
        className={`${
          small ? styles[buttonMobileClass] : styles[buttonClass]
        } ${className && className} ${styles.inputBtn} ${svg &&
          styles.inputWithIcon}`}
        type="submit"
        onClick={onClick}
        value={value}
      />
    </div>
  );
};

ButtonInput.propTypes = {
  onClick: PropTypes.func,
  value: PropTypes.string,
  color: PropTypes.oneOf(['Orange', 'White']),
  className: PropTypes.string,
  svg: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  small: PropTypes.bool
};

ButtonInput.defaultProps = {
  color: 'White',
  className: null,
  value: null,
  onClick: () => {},
  svg: null,
  small: false
};

export default ButtonInput;
