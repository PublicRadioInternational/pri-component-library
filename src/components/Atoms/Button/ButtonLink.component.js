/**
 * @file ButtonLink.component.js
 * Exports a button link component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styles from './Button.css';

/**
 * Component that renders a link button with a click handler.
 */
const ButtonLink = props => {
  const { url, onClick, className, children, color, small } = props;
  // Generate a class name based on the color.
  const buttonClass = `btn${color}`;
  const buttonMobileClass = `btnMobile${color}`;

  return (
    <a
      href={url}
      className={`${
        small ? styles[buttonMobileClass] : styles[buttonClass]
      } ${className && className}`}
      onClick={onClick}
    >
      <span
        className={`${styles.textLabel} ${small && styles.textLabelMobile}`}
      >
        {children}
      </span>
    </a>
  );
};

ButtonLink.propTypes = {
  url: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  color: PropTypes.oneOf(['Orange', 'White']),
  className: PropTypes.string,
  small: PropTypes.bool
};

ButtonLink.defaultProps = {
  url: null,
  color: 'White',
  className: null,
  children: null,
  onClick: () => {},
  small: false
};

export default ButtonLink;
