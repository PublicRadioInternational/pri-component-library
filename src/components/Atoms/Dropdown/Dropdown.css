/* import colors... */
@value colors: "../../00_global/colors.css";
@value blue from colors;
/* import fonts... */
@value fonts: "../../00_global/fonts.css";
@value copy from fonts;

/* Group */
.btnGroup {
  display: flex;
  position: relative;
  vertical-align: middle;
}

/* Dropdown */
.dropdown {
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid rgba(0, 0, 0, 0.15);
  border-radius: 0.25rem;
  display: none;
  left: 0;
  margin: 0.125rem 0 0;
  min-width: 10rem;
  padding: 0.5rem 0;
  position: absolute;
  top: 100%;
  z-index: 1;
}

.dropdownItem {
  clear: both;
  color: blue;
  display: block;
  font-size: 0.875em;
  font-weight: 400;
  padding: 0.25rem 1.5rem;
  width: 100%;
}
