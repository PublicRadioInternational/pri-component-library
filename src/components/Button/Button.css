.button {
  background-color: cyan;
  border: none;
  cursor: pointer;
  font-size: 1.1rem;
  font-weight: bold;
  padding: 0.5em 1em;
  text-transform: uppercase;
}

.button:hover {
  background-color: #eee;
}
