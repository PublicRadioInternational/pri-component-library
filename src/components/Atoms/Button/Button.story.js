/**
 * @file atoms.story.js
 * Story configuration for Atoms.
 */

import React from 'react';
import classNames from 'classnames/bind';
import { withInfo } from '@storybook/addon-info';
import { checkA11y } from '@storybook/addon-a11y';
import {
  withKnobs,
  text,
  select,
  boolean,
  object
} from '@storybook/addon-knobs/react';
import styles from '@sambego/storybook-styles';
import startCase from 'lodash/startCase';
import { shallow } from 'enzyme';
import expect from 'expect';
import {
  storiesOf,
  action,
  specs,
  describe,
  it,
  beforeEach,
  snapshot
} from '../../../../.storybook/facade';
import Button from './Button.component';
import btnStyles from './Button.css';

const cx = classNames.bind(btnStyles);

const stories = storiesOf('Atoms/Buttons', module);

// Convert flat colors array to object keyed by color for title cased label.
// Example: ['color'] to { color: 'Color' }.
const buttonColors = Button.colors.reduce(
  (obj, color) => ({ ...obj, [color.toLowerCase()]: startCase(color) }),
  {}
);

/**
 * Add storybook definition for buttons.
 */
stories.addDecorator(checkA11y);
stories.addDecorator(withKnobs);
stories.addDecorator(
  styles({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '100vh',
    backgroundColor: '#fff',
    padding: '0 1rem'
  })
);

stories.add(
  'Info & Knobs',
  withInfo({
    inline: true,
    text: `
    Basic button UI element.
  `
  })(() => (
    <Button
      onClick={action('onClickHandler')}
      color={select('Color', buttonColors, Button.colors[0].toLowerCase())}
      disabled={boolean('Disabled')}
      small={boolean('Small')}
      href={text('URL')}
      type={select('Input Type', ['button', 'submit', 'reset'])}
      {...object('Other Attributes', {}, 'BTNOPT-1')}
    >
      {text('Label', 'Listen')}
    </Button>
  ))
);

const exampleSpecs = storiesOf('Atoms/Buttons/Examples');

exampleSpecs.addDecorator(checkA11y);
exampleSpecs.addDecorator(
  styles({
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    minHeight: '100vh',
    backgroundColor: '#fff',
    padding: '0 1rem'
  })
);

exampleSpecs.add('Default Listen Button', () => {
  const story = <Button onClick={action('button-clicked')}>Listen</Button>;

  specs(() =>
    describe('Default Listen Button', () => {
      let output; // eslint-disable-line

      beforeEach(() => {
        output = shallow(story);
      });

      it('Should default to <button> element.', () => {
        expect(output.type()).toEqual('button');
      });

      it('Should have "Listen" label.', () => {
        expect(output.text()).toEqual('Listen');
      });

      it('Should have class "btnWhite".', () => {
        expect(output.hasClass(cx('btnWhite'))).toEqual(true);
      });

      it('Should have type "button".', () => {
        expect(output.prop('type')).toEqual('button');
      });
    })
  );

  snapshot('Default Listen Button', story);

  return story;
});

exampleSpecs.add('Orange Donate Link', () => {
  const story = (
    <Button onClick={action('button-clicked')} href="/donate" color="orange">
      Donate
    </Button>
  );

  specs(() =>
    describe('Orange Donate Button', () => {
      let output; // eslint-disable-line

      beforeEach(() => {
        output = shallow(story);
      });

      it('Should convert to <a> element.', () => {
        expect(output.type()).toEqual('a');
      });

      it('Should have "Donate" label.', () => {
        expect(output.text()).toEqual('Donate');
      });

      it('Should have class "btnOrange".', () => {
        expect(output.hasClass(cx('btnOrange'))).toEqual(true);
      });

      it('Should have href "/donate".', () => {
        expect(output.prop('href')).toEqual('/donate');
      });
    })
  );

  snapshot('Orange Donate Button', story);

  return story;
});

exampleSpecs.add('Blue Submit Button', () => {
  const story = (
    <Button onClick={action('button-clicked')} type="submit" color="blue">
      Submit
    </Button>
  );

  specs(() =>
    describe('Blue Submit Button', () => {
      let output; // eslint-disable-line

      beforeEach(() => {
        output = shallow(story);
      });

      it('Should default to <button> element.', () => {
        expect(output.type()).toEqual('button');
      });

      it('Should have "Submit" label.', () => {
        expect(output.text()).toEqual('Submit');
      });

      it('Should have class "btnBlue".', () => {
        expect(output.hasClass(cx('btnBlue'))).toEqual(true);
      });

      it('Should have type "submit".', () => {
        expect(output.prop('type')).toEqual('submit');
      });
    })
  );

  snapshot('Blue Submit Button', story);

  return story;
});

exampleSpecs.add('Dark Reset Input', () => {
  const story = (
    <Button
      onClick={action('button-clicked')}
      component="input"
      type="reset"
      color="dark"
      value="Reset"
    />
  );

  specs(() =>
    describe('Blue Submit Button', () => {
      let output; // eslint-disable-line

      beforeEach(() => {
        output = shallow(story);
      });

      it('Should convert to <input> element.', () => {
        expect(output.type()).toEqual('input');
      });

      it('Should have "Reset" label (as value.)', () => {
        expect(output.prop('value')).toEqual('Reset');
      });

      it('Should have class "btnDark".', () => {
        expect(output.hasClass(cx('btnDark'))).toEqual(true);
      });

      it('Should have type "reset".', () => {
        expect(output.prop('type')).toEqual('reset');
      });
    })
  );

  snapshot('Blue Submit Button', story);

  return story;
});
