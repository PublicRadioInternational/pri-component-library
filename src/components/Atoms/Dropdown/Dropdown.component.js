/**
 * @file Dropdown.component.js
 * Creates dropdown component.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styles from './Dropdown.css';
import Button from '../Button/Button.component';

// const Dropdown = ({ items, btnTitle, isToggled }) => (
class Dropdown extends React.Component {
  constructor(props) {
    super(props);
    this.handleToggleChange = this.handleToggleChange.bind(this);
    this.state = { toggled: false };
  }
  handleToggleChange() {
    this.setState(prevState => ({
      toggled: !prevState.toggled
    }));
  }
  render() {
    return (
      <div className={styles.btnGroup}>
        <Button
          hidden={false}
          title={this.props.btnTitle}
          url="#"
          className="btnGrpOrange"
          ariaHaspopup
          ariaExpanded={false}
        />
        <Button
          button
          hidden
          title="Toggle Dropdown"
          url="#"
          className="btnDropdownOrange"
          ariaHaspopup
          ariaExpanded={false}
          toggled={this.handleToggleChange}
        />
        <div
          className={styles.dropdown}
          data-toggled={this.state.toggled ? 'toggled' : null}
        >
          {this.props.items.map(item => (
            <a className={styles.dropdownItem} href={item.url} key={item.id}>
              {item.title}
            </a>
          ))}
        </div>
      </div>
    );
  }
}

Dropdown.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  btnTitle: PropTypes.string.isRequired
};

export default Dropdown;
