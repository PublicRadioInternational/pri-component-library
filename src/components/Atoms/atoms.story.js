/**
 * @file atoms.story.js
 * Story configuration for Atoms.
 */

import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Button from './Button/Button.component';
import Dropdown from './Dropdown/Dropdown.component';

storiesOf('Atoms/Buttons', module).add('Default', () => (
  <Button
    onClick={action('button-clicked')}
    title="Listen"
    url="#"
    dropdown={false}
    className="btn"
  />
));

storiesOf('Atoms/Buttons', module).add('Orange', () => (
  <Button
    onClick={action('button-clicked')}
    title="Donate"
    url="#"
    className="btnOrange"
  />
));

storiesOf('Atoms/Dropdown', module).add('Default', () => (
  <Dropdown
    btnTitle="Listen"
    items={[
      {
        id: 1,
        url: 'http://google.com',
        title: 'Google.com'
      },
      {
        id: 2,
        url: '/',
        title: 'Local url'
      }
    ]}
  />
));
