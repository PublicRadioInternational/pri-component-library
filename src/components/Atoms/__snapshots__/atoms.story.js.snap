// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blue Submit Button 1`] = `
<button
  className="btnBlue"
  disabled={false}
  onClick={undefined}
  type="submit"
>
  Submit
</button>
`;

exports[`Blue Submit Button 2`] = `
<button
  className="btnBlue"
  disabled={false}
  onClick={undefined}
  type="submit"
>
  Submit
</button>
`;

exports[`Blue Submit Button 3`] = `
<input
  className="btnDark"
  disabled={false}
  onClick={undefined}
  type="reset"
  value="Reset"
/>
`;

exports[`Dark Reset Input 1`] = `
<input
  className="btnDark"
  disabled={false}
  onClick={undefined}
  type="reset"
  value="Reset"
/>
`;

exports[`Default Listen Button 1`] = `
<button
  className="btnWhite"
  disabled={false}
  onClick={undefined}
  type="button"
>
  Listen
</button>
`;

exports[`Default Listen Button 2`] = `
<button
  className="btnWhite"
  disabled={false}
  onClick={undefined}
  type="button"
>
  Listen
</button>
`;

exports[`Orange Donate Button 1`] = `
<a
  className="btnOrange"
  href="/donate"
  onClick={undefined}
  role="button"
>
  Donate
</a>
`;

exports[`Orange Donate Link 1`] = `
<a
  className="btnOrange"
  href="/donate"
  onClick={undefined}
  role="button"
>
  Donate
</a>
`;

exports[`With Knobs 1`] = `
<button
  className="btnWhite"
  disabled={false}
  href={undefined}
  onClick={undefined}
  type="button"
>
  Listen
</button>
`;
